File: packages/types/src/iep.ts
// Core IEP Interface
import type { DataPoint } from './common';

export interface IEP {
  id: string;
  learnerId: string;
  schoolYear: string;
  startDate: Date;
  endDate: Date;
  reviewDates: Date[];
  nextReviewDate: Date;
  status: IEPStatus;
  goals: IEPGoal[];
  accommodations: Accommodation[];
  modifications: Modification[];
  services: Service[];
  team: IEPTeamMember[];
  meetingNotes: MeetingNote[];
  documents: IEPDocument[];
  createdBy: string;
  createdAt: Date;
  updatedAt: Date;
  lastReviewedAt?: Date;
}

export type IEPStatus = 'draft' | 'active' | 'review-due' | 'under-review' | 'expired' | 'archived';

// IEP Goal
export interface IEPGoal {
  id: string;
  iepId: string;
  domain: 'reading' | 'math' | 'speech' | 'writing' | 'social-emotional' | 'motor-skills' | 'life-skills' | 'behavioral' | 'other';
  title: string;
  description: string;
  measurableObjective: string;
  baselineLevel: string;
  targetLevel: string;
  targetDate: Date;
  progress: number; // 0-100
  status: 'not-started' | 'in-progress' | 'mastered' | 'discontinued' | 'modified';
  strategies: string[];
  assessmentMethod: string;
  dataCollectionFrequency: string;
  progressReports: GoalProgressReport[];
  standardsAligned?: string[]; // Educational standards
  createdAt: Date;
  updatedAt: Date;
}

// Goal Progress Report
export interface GoalProgressReport {
  id: string;
  goalId: string;
  reportedBy: string;
  date: Date;
  progressPercentage: number;
  observations: string;
  dataPoints?: DataPoint[];
  nextSteps?: string;
}

// Accommodation
export interface Accommodation {
  id: string;
  iepId: string;
  category: 'instructional' | 'assessment' | 'environmental' | 'behavioral';
  type: 'presentation' | 'response' | 'setting' | 'timing' | 'scheduling';
  description: string;
  frequency: 'always' | 'as-needed' | 'specific-subjects' | 'specific-activities';
  subjects?: string[];
  implementationNotes?: string;
  isActive: boolean;
}

// Modification
export interface Modification {
  id: string;
  iepId: string;
  subject: string;
  description: string;
  type: 'curriculum' | 'assignment' | 'grading' | 'other';
  implementationDetails: string;
  isActive: boolean;
}

// Service
export interface Service {
  id: string;
  iepId: string;
  type: ServiceType;
  description: string;
  provider: string;
  providerId?: string;
  frequency: string; // e.g., "2x per week"
  duration: number; // minutes per session
  location: string;
  schedule?: string; // e.g., "Monday & Wednesday, 10:00 AM"
  startDate: Date;
  endDate?: Date;
  isActive: boolean;
  sessionLogs?: ServiceLog[];
}

export type ServiceType = 
  | 'speech-language-therapy'
  | 'occupational-therapy'
  | 'physical-therapy'
  | 'counseling'
  | 'behavioral-support'
  | 'tutoring'
  | 'adapted-pe'
  | 'assistive-technology'
  | 'vision-support'
  | 'hearing-support'
  | 'nursing'
  | 'transportation'
  | 'other';

// Service Log
export interface ServiceLog {
  id: string;
  serviceId: string;
  date: Date;
  duration: number; // actual minutes
  attendance: 'present' | 'absent' | 'partial';
  activitiesConducted: string;
  progressNotes: string;
  providerId: string;
  createdAt: Date;
}

// IEP Team Member
export interface IEPTeamMember {
  userId: string;
  role: 'parent' | 'teacher' | 'special-ed-teacher' | 'speech-therapist' | 'psychologist' | 'administrator' | 'counselor' | 'other';
  name: string;
  title?: string;
  email?: string;
  phone?: string;
  isPrimaryContact: boolean;
}

// Meeting Note
export interface MeetingNote {
  id: string;
  iepId: string;
  meetingType: 'initial' | 'annual' | 'quarterly' | 'amendment' | 'reevaluation' | 'other';
  date: Date;
  attendees: IEPTeamMember[];
  agenda?: string;
  notes: string;
  decisions: string[];
  actionItems: ActionItem[];
  nextMeetingDate?: Date;
  createdBy: string;
  createdAt: Date;
}

export interface ActionItem {
  id: string;
  description: string;
  assignedTo: string;
  dueDate: Date;
  status: 'pending' | 'in-progress' | 'completed' | 'cancelled';
  completedAt?: Date;
}

// IEP Document
export interface IEPDocument {
  id: string;
  iepId: string;
  type: 'iep-form' | 'progress-report' | 'assessment' | 'consent-form' | 'meeting-notice' | 'evaluation' | 'other';
  title: string;
  filename: string;
  fileUrl: string;
  fileSize: number;
  mimeType: string;
  uploadedBy: string;
  uploadedAt: Date;
  isConfidential: boolean;
}

// IEP Timeline Event
export interface IEPTimelineEvent {
  id: string;
  iepId: string;
  eventType: 'created' | 'modified' | 'reviewed' | 'goal-added' | 'goal-completed' | 'service-added' | 'meeting-held' | 'document-uploaded';
  description: string;
  performedBy: string;
  timestamp: Date;
  metadata?: Record<string, any>;
}

// IEP Analytics
export interface IEPAnalytics {
  iepId: string;
  learnerId: string;
  totalGoals: number;
  completedGoals: number;
  goalsInProgress: number;
  averageGoalProgress: number; // 0-100
  goalsOnTrack: number;
  goalsBehindSchedule: number;
  daysUntilNextReview: number;
  complianceStatus: 'compliant' | 'needs-attention' | 'overdue';
  lastUpdated: Date;
}

File: apps/api-gateway-backend/app/schemas/iep.py
"""IEP schemas for request/response validation."""

from datetime import date, datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class IEPStatus(str, Enum):
    """IEP document status."""

    ACTIVE = "active"
    DRAFT = "draft"
    REVIEW_DUE = "review-due"
    ARCHIVED = "archived"


class IEPGoalStatus(str, Enum):
    """IEP goal status."""

    ON_TRACK = "on-track"
    NEEDS_ATTENTION = "needs-attention"
    EXCEEDING = "exceeding"
    NOT_STARTED = "not-started"


class IEPGoalCreate(BaseModel):
    """Schema for creating IEP goal."""

    learner_id: str = Field(..., min_length=36, max_length=36)
    goal_name: str = Field(..., min_length=1, max_length=500)
    goal_description: Optional[str] = None
    category: str = Field(..., pattern=r"^(reading|math|social|motor|communication)$")
    current_level: str = Field(..., min_length=1, max_length=255)
    target_level: str = Field(..., min_length=1, max_length=255)
    start_date: date
    target_date: date
    accommodations: Optional[List[str]] = None


class IEPGoalUpdate(BaseModel):
    """Schema for updating IEP goal."""

    goal_name: Optional[str] = Field(None, min_length=1, max_length=500)
    goal_description: Optional[str] = None
    current_level: Optional[str] = None
    target_level: Optional[str] = None
    target_date: Optional[date] = None
    progress_percentage: Optional[int] = Field(None, ge=0, le=100)
    status: Optional[IEPGoalStatus] = None
    accommodations: Optional[List[str]] = None


class IEPDataPointCreate(BaseModel):
    """Schema for creating data point."""

    goal_id: str
    value: int = Field(..., ge=0, le=100)
    notes: Optional[str] = None
    recorded_by: str


class IEPDataPointResponse(BaseModel):
    """Schema for data point response."""

    id: str
    goal_id: str
    value: int
    notes: Optional[str] = None
    recorded_by: str
    created_at: datetime

    class Config:
        from_attributes = True


class IEPGoalResponse(BaseModel):
    """Schema for IEP goal response."""

    id: str
    learner_id: str
    goal_name: str
    goal_description: Optional[str] = None
    category: str
    current_level: str
    target_level: str
    start_date: date
    target_date: date
    progress_percentage: int
    status: IEPGoalStatus
    accommodations: Optional[List[str]] = None
    data_points: List[IEPDataPointResponse] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# IEP Document Schemas


class ServiceInfo(BaseModel):
    """Service information for IEP."""

    service: str
    frequency: str
    provider: str
    duration: Optional[str] = None


class IEPCreate(BaseModel):
    """Schema for creating IEP document."""

    learner_id: str = Field(..., min_length=36, max_length=36)
    case_manager: str = Field(..., min_length=1, max_length=255)
    date_created: date
    next_review: date
    effective_date: Optional[date] = None
    parent_contact: Optional[str] = Field(None, max_length=255)
    parent_phone: Optional[str] = Field(None, max_length=50)
    parent_email: Optional[str] = Field(None, max_length=255)
    services: Optional[List[ServiceInfo]] = []
    notes: Optional[str] = None


class IEPUpdate(BaseModel):
    """Schema for updating IEP document."""

    case_manager: Optional[str] = Field(None, min_length=1, max_length=255)
    next_review: Optional[date] = None
    status: Optional[IEPStatus] = None
    parent_contact: Optional[str] = Field(None, max_length=255)
    parent_phone: Optional[str] = Field(None, max_length=50)
    parent_email: Optional[str] = Field(None, max_length=255)
    services: Optional[List[ServiceInfo]] = None
    notes: Optional[str] = None


class IEPResponse(BaseModel):
    """Schema for IEP document response."""

    id: str
    learner_id: str
    teacher_id: str
    status: IEPStatus
    case_manager: str
    date_created: date
    last_modified: date
    next_review: date
    effective_date: Optional[date] = None
    parent_contact: Optional[str] = None
    parent_phone: Optional[str] = None
    parent_email: Optional[str] = None
    services: Optional[List[dict]] = []
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class IEPWithGoalsResponse(IEPResponse):
    """Schema for IEP with goals included."""

    goals: List[IEPGoalResponse] = []

    class Config:
        from_attributes = True
